esphome/components/remote_base/nec_protocol.cpp

```diff
@@ -6,97 +6,243 @@ namespace remote_base {
+      data.command |= mask;  // Logic '1'
     } else if (src.expect_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-      data.command &= ~mask;
+      // Logic '0', since the command is already initialized with 0, `data.command &= ~mask;` is not needed
     } else {
       return {};
     }
   }
 
-  while (src.peek_item(BIT_HIGH_US, BIT_ONE_LOW_US) || src.peek_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-    uint16_t command = 0;
-    for (uint16_t mask = 1; mask; mask <<= 1) {
-      if (src.expect_item(BIT_HIGH_US, BIT_ONE_LOW_US)) {
-        command |= mask;
-      } else if (src.expect_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-        command &= ~mask;
-      } else {
-        return {};
-      }
-    }
-
-    // Make sure the extra/repeated data matches original command
-    if (command != data.command) {
-      return {};
-    }
+  // Validate the final stop bit (end of the message frame)
+  if (!src.expect_mark(BIT_HIGH_US)) {
+    return {};
+  }
 
-    data.command_repeats += 1;
+  // Message frame received, `data.type = NECCodeType::FRAME_WITH_REPEATS` is already set
+  if (!NECProtocol::is_command_valid(data)) {
+    ESP_LOGW(TAG, "Decoded command invalid: 0x%04X", data.command);
   }
 
-  src.expect_mark(BIT_HIGH_US);
+  ESP_LOGVV(TAG, "Decoded %s, src_idx=%" PRIu32 ", src_sz=%" PRIi32,
+            this->get_protocol_type_and_fields_str(data).c_str(), src.get_index(), src.size());
+
   return data;
 }
+
 void NECProtocol::dump(const NECData &data) {
-  ESP_LOGI(TAG, "Received NEC: address=0x%04X, command=0x%04X command_repeats=%d", data.address, data.command,
-           data.command_repeats);
+  ESP_LOGI(TAG, "Received %s", this->get_protocol_type_and_fields_str(data).c_str());
+}
+
+std::string NECProtocol::get_protocol_type_and_fields_str(const NECData &data) {
+  std::string debug_message = "NEC ";
+  switch (data.type) {
+    case NECCodeType::FRAME_WITH_REPEATS:
+      debug_message += str_sprintf("Frame (%u-bit address)", NECProtocol::is_address_extended(data) ? 16 : 8);
+      break;
+    case NECCodeType::REPEATS_ONLY:
+      debug_message += "Repeat Code:";
+      break;
+    default:
+      debug_message += "Unknown";
+  }
+
+  if (data.type != NECCodeType::REPEATS_ONLY) {
+    debug_message += ": address=0x";
+    if (NECProtocol::is_address_extended(data)) {
+      debug_message += str_sprintf("%04X", data.address);
+    } else {
+      debug_message += str_sprintf("%02X, address#=0x%02X", data.address_bytes.lo, data.address_bytes.hi);
+    }
+
+    debug_message += str_sprintf(", command=0x%02X, command#=0x%02X, command_valid=%s,", data.command_bytes.lo,
+                                 data.command_bytes.hi, YESNO(NECProtocol::is_command_valid(data)));
+  }
+
+  debug_message += str_sprintf(" repeats=%u", data.repeats);
+
+  return debug_message;
+}
+
+void NECBinarySensor::dump_config() {
+  LOG_BINARY_SENSOR("", "Remote Receiver NEC Binary Sensor", this);
+  ESP_LOGCONFIG(TAG, "  Repeat Timeout: %u ms", this->repeat_timeout_ms_);
```

Comment:
Nitpicking a bit, but...No Reason For These To Be Title Case. haha
```suggestion
  LOG_BINARY_SENSOR("", "Remote receiver NEC binary sensor", this);
  ESP_LOGCONFIG(TAG, "  Repeat timeout: %u ms", this->repeat_timeout_ms_);
```

Reply:
That's what I wanted but then I decided to copy https://github.com/esphome/esphome/blob/58fe8b39b20c8fe9767107a6136ca30873b82056/esphome/components/remote_base/remote_base.cpp#L126 and stick with it. I agree that it looks ugly title case.